### ISSUE TITLE
chore(common): register .kmx / .kmp with MIME and with magic

### ISSUE THREAD

**Issue by srl295** [keymanapp/keyman#10967]
- [x] register .kmx (binary) and .kmp (zipfile) with MIME
- [X] register magic bytes in the appropriate place

.kmp will be challenging because the zip metadata is at the end

**Comment by srl295** [comment]
The form is here https://www.iana.org/assignments/media-types/media-types.xhtml#image

@mcdurdin some questions:

- I think you should perhaps have your email on the form. Still, I'm glad to get a draft going. In fact, the draft spec could live in the repo.
- Thinking of these: (or `vnd.sil`? )

| extension | mime                           | Spec Link  |
|-----------|--------------------------------|------------|
|   .kmx    | application/vnd.keyman.kmx     | [here](https://help.keyman.com/developer/current-version/reference/file-types/kmx) |
|   .kmp    | application/vnd.keyman.kmp+zip | [here](https://help.keyman.com/developer/current-version/reference/file-types/kmp) |

- also, update the pages at 'spec link' to actually link to a detailed specification of the format

**Comment by mcdurdin** [comment]
* KMX does not have formal documentation, should use KMXPlus documentation as a starting point. We need a new issue for this I think!

* KMXPlus documented here: https://github.com/keymanapp/keyman/blob/master/core/src/ldml/C7043_ldml.md, should we copy this to help repo, and give it a better name than 7043?

* KMP documentation includes links to metadata (https://help.keyman.com/developer/current-version/reference/file-types/metadata) which is most of the formal documentation we have.

vnd.keyman is appropriate. We have the keyman.com domain. We don't tend to use the name SIL in any internal names for Keyman, only in branding.

Happy to have my email on the form (perhaps my @keyman.com address).

**Comment by srl295** [comment]
- .kmp is going to be hard to detect.  There's no predictable order of the files. I don't see a lot of prior art for such zipfiles.
- .kmx detection works well.  See a patch https://github.com/file/file/commit/e36df1065e9a41d76301ab626bb361fe137e1eb0 - could I get a review of the text there and then I'l submit it to file?  It has the MIME type commented out for now but that can be added.

**Comment by srl295** [comment]
@mcdurdin suggested MIME template

```
KMX

Your Full Name: ≈

Your Email: ≈

Type Name: application

Subtype Name: 
 - Vendor Tree (vnd. prefix)
 - keyman.kmx

Required Parameters: N/A

Optional Parameters: N/A

Encoding Considerations: binary

Security Considerations:  Does not contain executable content. Application must validate that all data fields are within specification and reject invalid content. 
 ** Or just “the security issues associated with this type have not been assessed” ?

Interoperability Condsiderations: N/A

Published Specification: https://help.keyman.com/developer/current-version/reference/file-types/kmx

Application Usage: This format is produced by Keyman keyboard build tools, and consumed by Keyman keyboard software.

Fragment Identifier: N/A

Restrictions on usage: N/a

Provisional: No

Magic number:  At byte offset zero, “KXTS”

File Extension:  .kmx

Intended Usage: COMMON
```

**Comment by srl295** [comment]
and KMP
```
KMP

Your Full Name: ≈

Your Email: ≈

Type Name: application

Subtype Name: 
 - Vendor Tree (vnd. prefix)
 - keyman.kmp+zip

Required Parameters: N/A

Optional Parameters: N/A

Encoding Considerations: binary

Security Considerations:  As produced by Keyman tools, does not contain executable content. Application must validate that all data fields are within specification and reject invalid content.  As a zipfile, security considerations should be considered as described in https://www.iana.org/assignments/media-types/application/zip

Interoperability Condsiderations: N/A

Published Specification: https://help.keyman.com/developer/current-version/reference/file-types/kmp

Application Usage: This format is produced by Keyman keyboard build tools, and consumed by Keyman keyboard software.

Fragment Identifier: N/A

Restrictions on usage: N/a

Provisional: No

Magic Number: It's a zip file; thus commonly: PK\003\004

File Extension:  .kmp

Intended Usage: COMMON
```

**Comment by mcdurdin** [comment]
> * .kmp is going to be hard to detect.  There's no predictable order of the files. I don't see a lot of prior art for such zipfiles.

It looks like it is possible to scan from the end of the file with a negative offset. In which case, we read the EOCD record (we can assume the zip has no file comment), so the record will be 22 bytes long. Something like this seems to work (with a few safety checks along the way):

```
# Keyman Compiled Package File

# A zip file that contains a file named kmp.json
0   string  PK\003\004
# Check End of central directory record magic
>-22 string  PK\005\006
# Find start of central directory
>>(-6.l)   string   PK\001\002
# Scan central directory for string 'kmp.json', will suffice for a 
# package containing about 150 files
>>>(-6.l)   search/9000   kmp.json   Keyman Compiled Package File
!:ext kmp
```

I don't think we need to worry about kmp.inf, because that's an obsolete file format, and there will be very few .kmp files that don't have kmp.json.

Prior art on ZIP file content checks: https://github.com/file/file/blob/master/magic/Magdir/msooxml

**Comment by mcdurdin** [comment]
I think we should hold off on registering MIME types until we have a published spec for .kmx.

* Blocked by #11407

**Comment by mcdurdin** [comment]
I think we are unblocked now, what do we need to do to move this forward?

**Comment by mcdurdin** [comment]
MIME type applications submitted.

**Comment by mcdurdin** [comment]
Keyman mime types: https://github.com/file/file/commit/04edb32ba99ff23ff3969ee606f08dface4e1ff7

**Comment by mcdurdin** [comment]
* .kmp `application/vnd.keyman.kmp+zip` has now been registered.
* .kmx `application/vnd.keyman.kmx` is still in process.

We should add this information to documentation.

**Comment by mcdurdin** [comment]
Both are now registered:
* https://www.iana.org/assignments/media-types/application/vnd.keyman.kmx
* https://www.iana.org/assignments/media-types/application/vnd.keyman.kmp+zip

**Comment by srl295** [comment]
- [ ] remaining todo - add MIME type back to magic definition.

**Comment by srl295** [comment]
with the mime type installed:

```
$ TZ=UTC MAGIC=../magic/magic ./file ../tests/keyman-0.testfile  --mime-type
../tests/keyman-0.testfile: application/vnd.keyman.kmx
$ TZ=UTC MAGIC=../magic/magic ./file ../tests/keyman-1.testfile  --mime-type
../tests/keyman-1.testfile: application/vnd.keyman.kmx
$ TZ=UTC MAGIC=../magic/magic ./file ../tests/keyman-2.testfile  --mime-type
../tests/keyman-2.testfile: application/vnd.keyman.kmp+zip
```

**Comment by srl295** [comment]
mime type is now in  magic definition… done…

**Comment by mcdurdin** [comment]
Debian added in 12.0.0: 
* https://tracker.debian.org/news/1619936/accepted-media-types-1200-source-into-unstable/
* https://browse.dgit.debian.org/media-types.git/commit/?id=a3d3c45ee0f41c33eee63112d892021a86f1cad6

Ubuntu added in Questing (25.10): 
* https://git.launchpad.net/ubuntu/+source/media-types/commit/IANA/application.csv?h=ubuntu/questing&id=6da5b8cc91434bfb9aa20faec9db5f73acb34f44

**Comment by mcdurdin** [comment]
As far as I can tell, the mime-types database on Android hasn't really been updated (from Debian) since 2019. An update was committed in 2024:
* https://android.googlesource.com/platform/external/mime-support/+/c2e269cf1cbb30fb04f85c3c0c49590b2a7d2b80

But that update was [rolled back](https://android.googlesource.com/platform/external/mime-support/+/da8a6ce88ca6c7bd1f804bf1f655b60f43835a0c) referencing a private bug report and since then it appears that very few mime types have been added back.

Not sure if it is worth pursuing this -- I don't know that it will improve the Android support for .kmp anyway.
